packages/next-swc/crates/next-custom-transforms/src/transforms/page_static_info/collect_exported_const_visitor.rs

```diff
@@ -0,0 +1,208 @@
+use std::collections::{HashMap, HashSet};
+
+use serde_json::{Map, Number, Value};
+use turbopack_binding::swc::core::{
+    common::{pass::AstNodePath, Mark, SyntaxContext},
+    ecma::{
+        ast::{
+            BindingIdent, Decl, ExportDecl, Expr, Lit, Pat, Prop, PropName, PropOrSpread, VarDecl,
+            VarDeclKind, VarDeclarator,
+        },
+        utils::{ExprCtx, ExprExt},
+        visit::{AstParentNodeRef, VisitAstPath, VisitWithPath},
+    },
+};
+
+/// The values extracted for the corresponding AST node.
+/// refer extract_expored_const_values for the supported value types.
+/// Undefined / null is treated as None.
+pub enum Const {
+    Value(Value),
+    Unsupported(String),
+}
+
+pub(crate) struct CollectExportedConstVisitor {
+    pub properties: HashMap<String, Option<Const>>,
+    expr_ctx: ExprCtx,
+}
+
+impl CollectExportedConstVisitor {
+    pub fn new(properties_to_extract: HashSet<String>) -> Self {
+        Self {
+            properties: properties_to_extract
+                .into_iter()
+                .map(|p| (p, None))
+                .collect(),
+            expr_ctx: ExprCtx {
+                unresolved_ctxt: SyntaxContext::empty().apply_mark(Mark::new()),
+                is_unresolved_ref_safe: false,
+            },
+        }
+    }
+}
+
+impl VisitAstPath for CollectExportedConstVisitor {
+    fn visit_export_decl<'ast: 'r, 'r>(
```

Comment:
Since export declarations can only appear on top level, we don't need a Visitor here. You can just for each all ModuleDeclarations and pick the ones you are interested in. This would avoid walking the whole AST.

Also side note: In general avoid VisitAstPath in favor of Visit if you are not using the ast_path